File: src/panorama/fixtures/data/demo-org/acme-gateway/branches/add-retry-helper/files/internal/retry/retry.go
// Package retry gives the gateway a backoff loop for upstream calls.
package retry

import "time"

const maxAttempts = 3

// retryWithBackoff calls operation until it succeeds or the attempt budget is
// exhausted, doubling the delay between attempts.
func retryWithBackoff(operation func() error, attempts int) error {
	delay := 100 * time.Millisecond
	var last error
	for attempt := 1; attempt <= attempts; attempt++ {
		last = operation()
		if last == nil {
			return nil
		}
		if attempt < attempts {
			time.Sleep(delay)
			delay *= 2
		}
	}
	return last
}

// Do runs operation with the default attempt budget.
func Do(operation func() error) error {
	return retryWithBackoff(operation, maxAttempts)
}

File: src/panorama/fixtures/data/demo-org/acme-shared/base/src/retry.ts
const MAX_ATTEMPTS = 3;
const INITIAL_DELAY_MS = 100;

// The organisation's retry policy: bounded attempts with exponential backoff.
// Services calling each other should import this rather than writing their own
// loop, so timeout behaviour is uniform across the org.
export async function retryWithBackoff<T>(
  operation: () => Promise<T>,
  attempts: number = MAX_ATTEMPTS,
): Promise<T> {
  let delay = INITIAL_DELAY_MS;
  let last: unknown;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      return await operation();
    } catch (err) {
      last = err;
      if (attempt < attempts) {
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay *= 2;
      }
    }
  }
  throw last;
}
